connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala

```diff
@@ -189,94 +158,164 @@ object KustoWriter{
+    val buffer: BufferedWriter = new BufferedWriter(writer, GZIP_BUFFER_SIZE)
+    val csvWriter: CsvWriter = new CsvWriter(buffer, new CsvWriterSettings)
+    datasink.FileWriteResource(buffer, gzip, csvWriter, blob)
+  }
 
-    for (row <- rows) {
-      csvWriter.writeRow(csvSerializer.convertRow(row))
+  @throws[IOException]
+  private[kusto] def serializeRows(rows: Iterator[InternalRow],
+                                   storageUri: String,
+                                   parameters: KustoWriteResource): Seq[CloudBlockBlob] = {
+    import parameters._
+    val container: CloudBlobContainer = new CloudBlobContainer(new URI(storageUri))
+    val fileWriter: FileWriteResource = createFileWriter(schema, writeOptions.timeZone, coordinates, tmpTableName, storageUri, container)
+
+    val (_, _, blobs) = rows.foldLeft[(Long, FileWriteResource, Seq[CloudBlockBlob])]((0, fileWriter, Seq())) { case ((size, fileWriter, blobsCreated), row) =>
+      val formattedRow: CsvRowResult = convertRowToCSV(row, schema, writeOptions.timeZone)
+      val newTotalSize = size + formattedRow.rowByteSize
+      if (newTotalSize < GIGA_SIZE) {
+        fileWriter.csvWriter.writeRow(formattedRow.formattedRow)
+        (newTotalSize, fileWriter, blobsCreated)
+      } else {
+        finalizeFileWrite(fileWriter)
+        (0, createFileWriter(schema, writeOptions.timeZone, coordinates, tmpTableName, storageUri, container), blobsCreated :+ fileWriter.blob)
+      }
     }
 
-    buffer.flush()
-    gzip.flush()
+    if (blobs.isEmpty) {
+      finalizeFileWrite(fileWriter)
+      Seq(fileWriter.blob)
+    } else {
+      blobs
+    }
+  }
 
-    buffer.close()
-    gzip.close()
+  def finalizeFileWrite(fileWriteResource: FileWriteResource): Unit = {
+    fileWriteResource.buffer.flush()
+    fileWriteResource.gzip.flush()
+    fileWriteResource.buffer.close()
+    fileWriteResource.gzip.close()
   }
 
-  private def readIngestionResult(statusRecord: IngestionStatus): String = {
-    val objectMapper = new ObjectMapper()
-    val resultAsJson = objectMapper.writerWithDefaultPrettyPrinter.writeValueAsString (statusRecord)
+  def convertRowToCSV(row: InternalRow, schema: StructType, timeZone: String): CsvRowResult = {
+    val dateFormat = FastDateFormat.getInstance("yyyy-MM-dd'T'HH:mm:ss.SSSXXX", TimeZone.getTimeZone(timeZone))
+    val schemaFields: Array[StructField] = schema.fields
 
-    resultAsJson
+    val (fields, size) = row.toSeq(schema).foldLeft[(Seq[String], Int)](Seq[String](), 0) { (res, curr) =>
+      val formattedField: String = curr match {
+        case DateType => DateTimeUtils.toJavaDate(row.getInt(res._1.size)).toString
+        case TimestampType => dateFormat.format(DateTimeUtils.toJavaTimestamp(row.getLong(res._1.size)))
+        case _ => row.get(res._1.size, schemaFields(res._1.size).dataType).toString
+
+      }
+      (res._1 :+ formattedField, res._2 + formattedField.getBytes(StandardCharsets.UTF_8).length)
```

Comment:
seems expensive - we should check performance changes

Reply:
I did some research, untill Scala 2.11 Seq had the same complexity like Vector in that manner. 
And now it has linear complexity for appending items. So after some research, **I think in that case the most efficient way will be to prepend a value to LIst**. 
Based on that article: [Performance Comparison between immutable Seq, List, Vector](https://danielasfregola.com/2015/06/15/which-immutable-scala-collection/)
And List/Seq/Vector data structure implementation.
What do you think?

Reply:
sorry for not being clear in my comments 
i was actually referring to the getBytes method . 
But it was an interesting read - in that manner it seems you are right